presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryLeakDetector.java

```diff
@@ -48,76 +45,49 @@
-     * @param queryMemoryReservations The memory reservations of queries in the GENERAL cluster memory pool.
-     */
-    void checkForMemoryLeaks(Supplier<List<BasicQueryInfo>> queryInfoSupplier, Map<QueryId, Long> queryMemoryReservations)
-    {
-        requireNonNull(queryInfoSupplier);
-        requireNonNull(queryMemoryReservations);
-
-        Map<QueryId, BasicQueryInfo> queryIdToInfo = Maps.uniqueIndex(queryInfoSupplier.get(), BasicQueryInfo::getQueryId);
-
-        Map<QueryId, Long> leakedQueryReservations = queryMemoryReservations.entrySet()
-                .stream()
-                .filter(entry -> entry.getValue() > 0)
-                .filter(entry -> isLeaked(queryIdToInfo, entry.getKey()))
-                .collect(toImmutableMap(Entry::getKey, Entry::getValue));
-
-        if (!leakedQueryReservations.isEmpty()) {
-            log.debug("Memory leak detected. The following queries are already finished, " +
-                    "but they have memory reservations on some worker node(s): %s", leakedQueryReservations);
-        }
-
-        synchronized (this) {
-            leakedQueries = ImmutableSet.copyOf(leakedQueryReservations.keySet());
-        }
-    }
+    @GuardedBy("this")
+    private long leakedBytes;
 
     /**
-     *
-     * @param runningQueriesSupplier All running queries on a cluster.
+     * @param queryIdToInfo All queries that the coordinator knows about, along with their optional query info.
      * @param queryMemoryReservations The memory reservations of queries in the GENERAL cluster memory pool.
      */
-    void checkForClusterMemoryLeaks(Supplier<Optional<List<QueryId>>> runningQueriesSupplier, Map<QueryId, Long> queryMemoryReservations)
+    void checkForMemoryLeaks(Map<QueryId, Optional<BasicQueryInfo>> queryIdToInfo, Map<QueryId, Long> queryMemoryReservations)
     {
-        requireNonNull(runningQueriesSupplier);
-        requireNonNull(queryMemoryReservations);
-
-        Optional<List<QueryId>> runningQueries = runningQueriesSupplier.get();
-
         Map<QueryId, Long> leakedQueryReservations = queryMemoryReservations.entrySet()
                 .stream()
                 .filter(entry -> entry.getValue() > 0)
-                .filter(entry -> !runningQueries.isPresent() || !runningQueries.get().contains(entry.getKey()))
+                .filter(entry -> isLeaked(queryIdToInfo, entry.getKey()))
                 .collect(toImmutableMap(Entry::getKey, Entry::getValue));
 
+        long leakedBytesThisTime = leakedQueryReservations.values().stream().reduce(0L, Long::sum);
         if (!leakedQueryReservations.isEmpty()) {
-            log.debug("Memory leak detected. The following queries are already finished, " +
-                    "but they have memory reservations on some worker node(s): %s", leakedQueryReservations);
+            log.warn("Memory leak of %s detected. The following queries are already finished, " +
```

Comment:
Is it intentional to change this to warn?

Reply:
Yes. The idea is to reduce the frequency of this and make the logging less silent. IMHO, leaks should be brought out in logs. In fact, in my previous company, we made this logging WARN internally so it was that useful :-P